### ISSUE TITLE
JIT config breaks ephemeral runners for GHES

### ISSUE THREAD

**Issue by jkruse14** [philips-labs/terraform-aws-github-runner#3392]
#3350 breaks ephemeral runners. When running with `enable_ephemeral_runners = true`, I get:

```
<13>Jul 26 23:57:02 user-data: Starting runner after 0:1:32
<13>Jul 26 23:57:02 user-data: Starting the runner as user ec2-user
<13>Jul 26 23:57:02 user-data: Starting the runner in ephemeral mode
[   92.500730] cloud-init[2611]: ci-info: no authorized ssh keys fingerprints found for user ec2-user.
ci-info: no authorized ssh keys fingerprints found for user ec2-user.
<snip>
[   92.562938] cloud-init[2611]: Cloud-init v. 19.3-46.amzn2.0.1 finished at Wed, 26 Jul 2023 23:57:02 +0000. Datasource DataSourceEc2.  Up 92.55 seconds
<13>Jul 26 23:57:03 user-data: Unrecognized command-line input arguments for command run: 'ephemeral, url, token, runnergroup'. For usage refer to: .\config.cmd --help or ./config.sh --help
<13>Jul 26 23:57:03 user-data: An error occurred: Not configured. Run config.(sh/cmd) to configure the runner.
<13>Jul 26 23:57:03 user-data: Runner listener exit with terminated error, stop the service, no retry needed.
<13>Jul 26 23:57:03 user-data: Exiting runner...
```

Notice:
**user-data: Unrecognized command-line input arguments for command run: 'ephemeral, url, token, runnergroup'. For usage refer to: .\config.cmd --help or ./config.sh --help
**

setting `enable_ephemeral_runners = false` allows it to work. The difference is non-ephemeral runners do not use jitconfig

The code in question is the template for start_runner.sh, line 113:

`sudo --preserve-env=RUNNER_ALLOW_RUNASROOT -u "$run_as" -- ./run.sh --jitconfig $${config}`

Also of note, I'm using the multi-runner config.

**Comment by GuptaNavdeep1983** [comment]
@jkruse14 can you please confirm the action runner version used by your deployment?

**Comment by strophy** [comment]
Hi @jkruse14 this looks like the same issue I am seeing here: https://github.com/philips-labs/terraform-aws-github-runner/issues/3390#issuecomment-1651398112

I have enabled a lot more logging and found that the plaintext `jitconfig` is replaced in the `$${config}` variable, but I think based on the [documentation](https://docs.github.com/en/actions/security-guides/security-hardening-for-github-actions#using-just-in-time-runners) that this should be encoded in some way?

**Comment by strophy** [comment]
@jkruse14 I have the same issue, runner version `2.306.0`

**Comment by GuptaNavdeep1983** [comment]
@strophy, can you check the value of the SSM parameter for jit config? This value is fetched in encoded form from GitHub api call and should be stored in SSM parameter in same form.

**Comment by strophy** [comment]
@GuptaNavdeep1983 yes, I think that is where we are going wrong. The value is stored in plaintext form in SSM Parameter Store like this:

key: `/github-action-runners/multi-runner/linux-x64-platform/runners/tokens/i-0017b273c05245c3f`
value: `--url https://github.com/strophy/platform --token A7BRI<REDACTED>5XA --ephemeral`

**Comment by GuptaNavdeep1983** [comment]
@strophy, Can you also confirm if you are using repo level runners or org level runners?

**Comment by strophy** [comment]
@GuptaNavdeep1983 this is a config for repo-level runners

**Comment by strophy** [comment]
Can you let me know how to enable the debug log output here: https://github.com/philips-labs/terraform-aws-github-runner/blob/main/lambdas/functions/control-plane/src/scale-runners/scale-up.ts#L388-L390

Setting `log_level = "debug"` in `examples/multi-runner/main.tf` doesn't seem to enable this particular log output from the lambda in the `/aws/lambda/multi-runner-linux-x64-platform-scale-up` log group

**Comment by GuptaNavdeep1983** [comment]
You can directly enable that in scale-up lambda function environment variable if this doesnt work out. I still need to check why its not enabled by setting this variable.

**Comment by strophy** [comment]
@GuptaNavdeep1983 that helped me figure it out, it was user error all along! I was thinking that I would need to rebuild the lambdas in order to enable that debug, and then realized that we did not download the pre-built lambdas for v4.0.0. The lambdas were still deployed with v3.6.1, resulting in plaintext vars being written to Parameter Store. Downloading the v4.0.0 lambdas resolved the issue and encoded text is now written to Parameter Store. Sorry about that, and thanks for your very helpful and prompt support!

**Comment by GuptaNavdeep1983** [comment]
I hope @jkruse14 can also confirm if its same for him and we can close this issue accordingly.

**Comment by mariusfilipowski** [comment]
Still this issue will arise with GitHub Enterprise Server since it does not have this API, correct?

**Comment by npalm** [comment]
@mariusfilipowski we have only tested the change on GitHub Cloud. Not on enterprise. I guess you are right that the API is not available yet. Which means we have to patch the code. Maybe we should add a variable to make the explicit choice to enable / disable JIT config. And only put it on by default if ephemeral is set to true.

**Comment by mariusfilipowski** [comment]
Thank you. That sounds like a good approach.

**Comment by npalm** [comment]
@mariusfilipowski would you have time to check / test the chnages in #3393 

- Ensure you build the lambda's
- Disable jit config by setting `enable_jit_config = false`

**Comment by jkruse14** [comment]
bah... I forgot about the lambdas! Yes, that was my issue! I've updated them and everything is working now.

Is it possible to have lambdas automatically upgraded similar to how the runner application is upgraded? Have an S3 bucket with the binaries on your side, and a lambda in the client module to check for updates? It could sync between the client bucket and your bucket if it sees there has been an updated file. This single syncer would likely need fewer updates than the three current lambdas and would make it less likely we have issues like, where a secondary process/step is missed.

**Comment by npalm** [comment]
> bah... I forgot about the lambdas! Yes, that was my issue! I've updated them and everything is working now.
> 
> Is it possible to have lambdas automatically upgraded similar to how the runner application is upgraded? Have an S3 bucket with the binaries on your side, and a lambda in the client module to check for updates? It could sync between the client bucket and your bucket if it sees there has been an updated file. This single syncer would likely need fewer updates than the three current lambdas and would make it less likely we have issues like, where a secondary process/step is missed.

Good point, we still looking for options to publish the lambd'as but having some limitation (process wise).

**Comment by npalm** [comment]
Re-open the issue till we have #3393 merged

**Comment by npalm** [comment]
We also figured out, or have to say not figured out that we are not sure what the auto update behavior is for JIT config. For JIT config you cannot pass this option. For this reason we will disable JIT config by default. And let the user explicit enable the feature.

**Comment by mariusfilipowski** [comment]
> @mariusfilipowski would you have time to check / test the chnages in #3393
> 
> * Ensure you build the lambda's
> * Disable jit config by setting `enable_jit_config = false`

Sadly, I have no workflow to build the lambdas and test a new version in a simple way. This would be quite complicated. So I have to deny. Perhaps in the future we can simplify something on our side to test such changes.
